bindings/rascal/models/krr.py

```diff
@@ -199,11 +204,43 @@ def get_representation_calculator(self):
         return self.kernel._rep
 
 
+def get_strides(frames):
+    Nstructures = len(frames)
+    Ngrad_stride = [0]
+    Ngrads = 0
+    for frame in frames:
+        n_at = len(frame)
+        Ngrad_stride.append(n_at * 3)
+        Ngrads += n_at * 3
+    Ngrad_stride = np.cumsum(Ngrad_stride) + Nstructures
+    return Nstructures, Ngrads, Ngrad_stride
+
+
+def compute(i_frame, frame, representation, X_sparse, kernel):
+    feat = representation.transform([frame])
+    en_row = kernel(feat, X_sparse)
+    grad_rows = kernel(feat, X_sparse, grad=(True, False))
+    return en_row, grad_rows
+
+
+def compute_KNM(frames, X_sparse, kernel, soap):
+    Nstructures, Ngrads, Ngrad_stride = get_strides(frames)
+    KNM = np.zeros((Nstructures + Ngrads, X_sparse.size()))
+    pbar = tqdm(frames, desc="compute KNM", leave=False)
```

Comment:
In fact, I'm not sure how I feel about including `tqdm` at such a deep level -- it seems wrong for the code to need to know about something that's in principle unrelated, but I'm not sure I see a better solution.

One idea would instead be to wrap `frames` in a `tqdm` _before_ passing it to this function in whatever notebook uses these - will look into this later.

Reply:
> will look into this later.

Did you try this? I would find it much more elegant, and it gives the user more control over the tqdm display name/positions/style etc.

Reply:
Not yet; I ended up just using the fix in cf0e660.

Reply:
Ok, I found a way that works -- it's not pretty, because the function iterates through the frames twice and I can't find a way to reset a tqdm bar once it's finished (you'd think that would be a pretty basic feature, but whatever).